API: make nvim_win_set_option() set the window value

bfredl opened the pull request:
Compare these two behaviors:
```
e file1
set winhl=Normal:ErrorMsg
e file2
```
and
```
e file1
setlocal winhl=Normal:ErrorMsg
e file2
```
In the former case the option is set for the window, and used for both files, but in the later case, the option is reset when switich buffer (buffer-local value).

`nvim_win_set_option(0, 'winhl', 'Normal:ErrorMsg')` should arguably work like the former and set the option for the entire window. We could also allow `nvim_buf_set_option(0, 'winhl', 'Normal:ErrorMsg')`. If the buffer is shown in a window this would work like `setl` in that window. If it is currently not shown in a window, it will set the value used when the buffer is visited in a new window.

bfredl committed:
API: make nvim_win_set_option() set window-global, not buffer-local value

bfredl commented:
Ref discussion https://github.com/neovim/neovim/pull/9100#issuecomment-429125151 and below.

bfredl commented:
NB: the `!(flags & SOPT_GLOBAL)` exception is for `statusline`. Because `:set` for it will set the global value for _all_ windows, so `:setl` is the best we can do there. This is a one-of-a-kind option that doesn't work like any other option.

justinmk commented:
> NB: the `!(flags & SOPT_GLOBAL)` exception is for `statusline`.

Probably should be mentioned in a source code comment.

> This is a one-of-a-kind option that doesn't work like any other option.

Sigh :)

jamessan commented:
> > This is a one-of-a-kind option that doesn't work like any other option.
> 
> Sigh :)

That just means we need to create some more options like that, so it isn't so lonely. 👼

bfredl commented:
Rebased and added tests. 

`eq({false, "Failed to get value for option 'statusline'"}, meth_pcall(curwin, 'get_option', 'statusline'))` is consistent with master, but perhaps not desirable (global-local option with unset local value).
